forgejo-mirror fails after every reboot: unattended SSH auth tied to FIDO key + locked keyring

## Summary

`forgejo-mirror.service` (hourly push-mirror of `main` → private Forgejo ledger, ADR-037/038)
authenticates to the local Forgejo over SSH using a **FIDO2 hardware key**
(`~/.ssh/id_ed25519_sk_htpc_signing`, an `ED25519-SK` key) served by the `gcr`
(gnome-keyring) ssh-agent. This works fine while the login keyring is unlocked, but on the
**first run after a host reboot** the keyring is locked, so the agent pops an interactive
prompt (keyring password + YubiKey touch). For an unattended systemd timer this is wrong:
if no one approves it, the mirror fails — and then keeps failing every hour until the
keyring is unlocked by some other means.

This bit us on the 2026-06-23 update+reboot: the operator saw an unexplained
password+touch prompt, denied it (couldn't tell what wanted the key), and the mirror has
been failing hourly ever since.

## Evidence (boot `89e5b547…`, 2026-06-23, boot at 16:34:29 CEST)

- 18 consecutive **successful** unattended runs before the reboot: every hour from
  `22:00:19` (Jun-22) through `15:48:45` (Jun-23), each finishing in ~1–1.3 s. So the key
  is effectively no-touch in steady state (the keyring was already unlocked).
- First post-reboot run **failed** at `16:46:58`, the same second `gcr-ssh-agent.service`
  (re)started (`16:46:59`):
  ```
  sign_and_send_pubkey: signing failed for ED25519-SK ".../id_ed25519_sk_htpc_signing" from agent: agent refused operation
  git@127.0.0.1: Permission denied (publickey).
  [ERROR] fetch failed (origin or forgejo unreachable)
  ```
- Next run `17:51:28` **also failed** (44 s hang → `[ERROR] fetch failed (origin or forgejo
  unreachable)`). Keyring still locked. `systemctl --user is-active forgejo-mirror` = `failed`.
- Root cause of the *lock*: this is a headless HTPC; the operator logs in on **tty2
  (console)**, which does **not** auto-unlock gnome-keyring the way a gdm graphical login
  does. So the keyring stays locked after every reboot until something unlocks it.

## Why this is the wrong design (not just a one-off)

Using the **commit-signing** FIDO key for **localhost transport auth** conflates two
concerns. The signing key's job is to attest provenance on `main` (the integrity gate). The
mirror's git push to `git@127.0.0.1` only needs "can I push to a same-host container." Tying
that hourly unattended job to an interactive keyring + hardware-presence path guarantees a
post-reboot failure + recurring `ForgejoMirrorFailed` noise every time the box restarts.

Also: the script reports an **auth** failure as `fetch failed (origin or forgejo
unreachable)` — misleading; reachability was fine, the agent refused to sign.

## Immediate remediation (to clear the current failure)

1. From the tty2 console, unlock the keyring once (e.g. `ssh git@127.0.0.1` and approve, or
   any keyring-touching app), then `systemctl --user start forgejo-mirror.service` and
   confirm it reaches `Finished`.

## Proposed durable fix

Decouple the unattended mirror from the interactive FIDO/keyring path:

- Generate a dedicated, passphrase-less `ed25519` **deploy key** for the mirror, add it to
  the Forgejo `patriark/homelab` repo as a push key, and pin it for the localhost remote via
  `~/.ssh/config`:
  ```
  Host 127.0.0.1
      IdentityFile ~/.ssh/id_ed25519_forgejo_mirror
      IdentitiesOnly yes
  ```
  Exposure is loopback-only to a same-host container — acceptable for an unattended job.
  Commit signing stays on the FIDO key (unchanged).
- Fix the misleading error string in `scripts/mirror-to-forgejo.sh` (distinguish auth vs
  reachability).
- Consider: should the *first* post-reboot mirror failure suppress `ForgejoMirrorFailed`
  until N consecutive failures, so a transient post-reboot lock isn't a tamper-canary
  false positive? (Keep the canary semantics for non-fast-forward / bad-signature cases.)

Alternative if the operator prefers all-hardware: configure PAM/systemd to auto-unlock the
login keyring at boot AND confirm the key is `no-touch-required` — but this re-introduces an
unlock dependency and is more fragile than a scoped deploy key.

## Acceptance criteria

- [ ] Mirror completes unattended on the first run after a cold reboot (no prompt, no failure).
- [ ] No interactive YubiKey/keyring prompt is ever triggered by the timer.
- [ ] Error messages distinguish auth failure from unreachability.
- [ ] Commit-signing provenance on `main` is unchanged (still the FIDO key).

## References

- `scripts/mirror-to-forgejo.sh`, `forgejo-mirror.{service,timer}`
- `~/.ssh/id_ed25519_sk_htpc_signing` (ED25519-SK)
- ADR-037 (forge center-of-gravity), ADR-038 (merge-commit-only), CLAUDE.md "Forgejo ledger"
- Memory: `feedback_commit_signing_via_github_squash` (FIDO key, SSH_AUTH_SOCK quirks)

--- a/scripts/mirror-to-forgejo.sh
+++ b/scripts/mirror-to-forgejo.sh
@@ -32,6 +32,19 @@ REMOTE="${MIRROR_REMOTE:-forgejo}"
 METRICS_DIR="${HOME}/containers/data/backup-metrics"
 METRICS_FILE="${METRICS_FILE:-${METRICS_DIR}/forgejo-mirror.prom}"
 
+# SSH identity for the Forgejo loopback push (ADR-032). A DEDICATED, non-FIDO
+# ed25519 deploy key — NOT the FIDO signing key. Forgejo 15.0.3 carries the
+# go1.26.4 fix for CVE-2026-39831: x/crypto/ssh now enforces FIDO User-Presence,
+# so the no-touch signing key's touch-less auth signatures are rejected mid-auth
+# (the connection drops right after it sends the signature) and the push fails.
+# A plain ed25519 key has no User-Presence concept, so it authenticates
+# unattended on the patched server. `-F none` keeps this independent of the
+# (chezmoi-managed) user SSH config so ONLY this one key is ever offered; the
+# explicit ssh:// remote URL supplies host/port/user. Commit signing is a
+# separate client-side path and is unaffected — the FIDO key still signs commits.
+MIRROR_SSH_KEY="${MIRROR_SSH_KEY:-$HOME/.ssh/id_ed25519_forgejo_mirror}"
+export GIT_SSH_COMMAND="ssh -F none -i $MIRROR_SSH_KEY -o IdentitiesOnly=yes -o IdentityAgent=none -o StrictHostKeyChecking=accept-new -o UserKnownHostsFile=$HOME/.ssh/known_hosts"
+
 log() { echo "[$(date '+%Y-%m-%d %H:%M:%S')] [$1] ${*:2}" >&2; }
 
 SUCCESS=0
@@ -68,6 +81,17 @@ finish() {
 trap finish EXIT
 
 cd "$REPO_ROOT" || { log ERROR "repo not found: $REPO_ROOT"; exit 1; }
+
+# Ensure the ledger remote is the explicit ssh:// URL, not the forgejo-local
+# ssh_config alias: with `-F none` in GIT_SSH_COMMAND above an alias would not
+# resolve, and the alias also carries the FIDO signing key (rejected post-CVE).
+# Idempotent; pins host/port/user for the dedicated deploy key.
+FORGEJO_URL="${FORGEJO_URL:-ssh://git@127.0.0.1:2222/patriark/homelab.git}"
+if [[ "$(git remote get-url "$REMOTE" 2>/dev/null)" != "$FORGEJO_URL" ]]; then
+    git remote set-url "$REMOTE" "$FORGEJO_URL" 2>/dev/null \
+        || git remote add "$REMOTE" "$FORGEJO_URL"
+    log INFO "pinned $REMOTE remote URL -> $FORGEJO_URL"
+fi
 
 if ! git fetch --quiet origin main || ! git fetch --quiet "$REMOTE" main; then
     log ERROR "fetch failed (origin or $REMOTE unreachable)"

